src/dbnode/network/server/tchannelthrift/node/service.go

```diff
@@ -729,153 +728,291 @@ func (s *service) FetchTagged(tctx thrift.Context, req *rpc.FetchTaggedRequest)
+type SegmentsIter interface {
+	BaseIter
+	// Current returns the current Segments. The result is only valid if Err() is nil.
+	Current() *rpc.Segments
+}
+
+type fetchTaggedResultsIter struct {
+	queryResults   map[index.ResultsMapHash]index.ResultsMapEntry
+	idResults      []IDResult
+	startInclusive time.Time
+	endExclusive   time.Time
+	db             storage.Database
+	idx            int
+	exhaustive     bool
+	fetchData      bool
+	cur            IDResult
+	err            error
+	nsID           ident.ID
+	docReader      *docs.EncodedDocumentReader
+	tagEncoder     serialize.TagEncoder
+	iOpts          instrument.Options
+}
+
+type fetchTaggedResultsIterOpts struct {
+	queryResult index.QueryResult
+	queryOpts   index.QueryOptions
+	fetchData   bool
+	db          storage.Database
+	docReader   *docs.EncodedDocumentReader
+	nsID        ident.ID
+	tagEncoder  serialize.TagEncoder
+	iOpts       instrument.Options
+}
+
+func newFetchTaggedResultsIter(opts fetchTaggedResultsIterOpts) FetchTaggedResultsIter { //nolint: gocritic
+	iter := &fetchTaggedResultsIter{
+		queryResults:   opts.queryResult.Results.Map().Iter(),
+		idResults:      make([]IDResult, 0, opts.queryResult.Results.Map().Len()),
+		exhaustive:     opts.queryResult.Exhaustive,
+		db:             opts.db,
+		fetchData:      opts.fetchData,
+		startInclusive: opts.queryOpts.StartInclusive,
+		endExclusive:   opts.queryOpts.EndExclusive,
+		nsID:           opts.nsID,
+		docReader:      opts.docReader,
+		tagEncoder:     opts.tagEncoder,
+		iOpts:          opts.iOpts,
 	}
-	defer sp.Finish()
 
-	i := 0
-	// Re-use reader and id for more memory-efficient processing of
-	// tags from doc.Metadata
-	reader := docs.NewEncodedDocumentReader()
-	for _, entry := range results.Map().Iter() {
-		idx := i
-		i++
+	return iter
+}
 
-		// NB(r): Use a bytes ID here so that this ID doesn't need to be
-		// copied by the blockRetriever in the streamRequest method when
-		// it checks if the ID is finalizeable or not with IsNoFinalize.
-		id := ident.BytesID(entry.Key())
+func (i *fetchTaggedResultsIter) NumIDs() int {
+	return len(i.queryResults)
+}
 
-		d := entry.Value()
-		metadata, err := docs.MetadataFromDocument(d, reader)
-		if err != nil {
-			return err
-		}
-		tags := idxconvert.ToSeriesTags(metadata, idxconvert.Opts{NoClone: true})
+func (i *fetchTaggedResultsIter) Exhaustive() bool {
+	return i.exhaustive
+}
+
+func (i *fetchTaggedResultsIter) Namespace() ident.ID {
+	return i.nsID
+}
 
-		enc := s.pools.tagEncoder.Get()
-		ctx.RegisterFinalizer(enc)
-		encodedTags, err := s.encodeTags(enc, tags)
-		if err != nil { // This is an invariant, should never happen
-			return tterrors.NewInternalError(err)
+func (i *fetchTaggedResultsIter) Next(ctx context.Context) bool {
+	if i.idx >= len(i.queryResults) {
+		return false
+	}
+	// TODO(rhall): don't request all series blocks at once.
+	if i.idx == 0 {
+		for _, entry := range i.queryResults { // nolint: gocritic
```

Comment:
If you want to avoid the `gocritic` comment altogether using array indexing actually bypasses stack usage here. (yes, basically performant Go starts becoming it's own dialect.. hah)

i.e.
```go
for i := range i.queryResults {
  result := IDResult{
    queryResult: i.queryResults[i],
    // ... rest

}
```